dhall/src/Dhall.hs

```diff
@@ -157,12 +165,47 @@ import qualified Dhall.Util
     This exception indicates that an invalid `Type` was provided to the `input`
     function
 -}
+type Extractor s a = Validation (ExtractErrors s a)
+
+newtype ExtractErrors s a =
+    ExtractErrors
+  { getErrors :: NonEmpty (ExtractError s a)
+  } deriving Semigroup
+
+typeError :: Expr s a -> Expr s a -> Extractor s a b
+typeError expected actual = Failure . ExtractErrors . (:| []) . TypeMismatch $ InvalidType expected actual
+
+extractError :: Text -> Extractor s a b
+extractError = Failure . ExtractErrors . (:| []) . ExtractError
+
+-- | opt-in monadic composition for `Extractor`. This is done this way to
+-- keep applicative behaviour on @ExtractErrors@
+validate :: Extractor s a b -> (b -> Extractor s a c) -> Extractor s a c
+validate (Success a) f = f a
+validate (Failure es) _ = Failure es
+
+instance (Pretty s, Pretty a, Typeable s, Typeable a) => Show (ExtractErrors s a) where
+    show (ExtractErrors es)  = unlines . Data.List.NonEmpty.toList . fmap show $ es
```

Comment:
Two suggestions for formatting the error output:

* If there is more than one error, prefix all of the errors with a brief message saying that there multiple errors
* Separate the errors visually (i.e. perhaps add a blank line between them if there's not one already)

Reply:
:+1: , i'll do that